// @gravity-ui/uikit/build/esm/components/TableColumnSetup/TableColumnSetup.js
'use client';
import React from 'react';
import { Gear } from '@gravity-ui/icons';
import { Button } from '../Button';
import { Icon } from '../Icon';
import { TableColumnSetup as NewTableColumnSetup } from '../Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup';
import { block } from '../utils/cn';
import i18n from './i18n';
import './TableColumnSetup.css';
const b = block('table-column-setup');
export const TableColumnSetup = (props) => {
    const { switcher, renderSwitcher: renderSwitcherProps, disabled, popupWidth, popupPlacement, className, items: propsItems, sortable = true, showStatus, onUpdate: propsOnUpdate, } = props;
    const renderStatus = () => {
        if (!showStatus) {
            return null;
        }
        const selected = propsItems.reduce((acc, cur) => (cur.selected ? acc + 1 : acc), 0);
        const all = propsItems.length;
        const status = `${selected}/${all}`;
        return React.createElement("span", { className: b('status') }, status);
    };
    const renderSwitcher = (switcherProps) => {
        return ((renderSwitcherProps === null || renderSwitcherProps === void 0 ? void 0 : renderSwitcherProps(switcherProps)) ||
            switcher || (React.createElement(Button, { disabled: disabled, onClick: switcherProps.onClick },
            React.createElement(Icon, { data: Gear }),
            i18n('button_switcher'),
            renderStatus())));
    };
    const items = propsItems.map(({ id, title, required, selected, sticky }) => ({
        id,
        title,
        isRequired: required,
        isSelected: selected,
        sticky,
    }));
    const onUpdate = (newSettings) => {
        propsOnUpdate(newSettings.map(({ id, isSelected }) => {
            const prevItem = propsItems.find((item) => item.id === id);
            return {
                id,
                selected: isSelected,
                title: prevItem === null || prevItem === void 0 ? void 0 : prevItem.title,
                required: prevItem === null || prevItem === void 0 ? void 0 : prevItem.required,
            };
        }));
    };
    return (React.createElement(NewTableColumnSetup, { items: items, onUpdate: onUpdate, popupPlacement: popupPlacement, popupWidth: popupWidth, renderSwitcher: renderSwitcher, sortable: sortable, className: b(null, className) }));
};
