// @gravity-ui/uikit/build/cjs/components/TreeSelect/TreeSelect.js
"use strict";
'use client';
Object.defineProperty(exports, "__esModule", { value: true });
exports.TreeSelect = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const hooks_1 = require("../../hooks");
const useOpenState_1 = require("../../hooks/useSelect/useOpenState");
const components_1 = require("../Select/components");
const SelectPopup_1 = require("../Select/components/SelectPopup/SelectPopup");
const TreeList_1 = require("../TreeList");
const mobile_1 = require("../mobile");
const useList_1 = require("../useList");
const cn_1 = require("../utils/cn");
const useControlledValue_1 = require("./hooks/useControlledValue");
require("./TreeSelect.css");
const b = (0, cn_1.block)('tree-select');
const defaultItemRenderer = (renderState) => {
    return react_1.default.createElement(useList_1.ListItemView, Object.assign({}, renderState.props, renderState.renderContainerProps));
};
exports.TreeSelect = react_1.default.forwardRef(function TreeSelect({ id, qa, title, placement, slotBeforeListBody, slotAfterListBody, size = 'm', defaultOpen, width, containerRef: propsContainerRef, className, containerClassName, popupClassName, open: propsOpen, multiple, popupWidth, popupDisablePortal, items, value: propsValue, defaultValue, placeholder, disabled = false, withExpandedState = true, defaultExpandedState = 'expanded', onClose, onOpenChange, onUpdate, renderControl, renderItem = defaultItemRenderer, renderContainer, mapItemDataToContentProps, onFocus, onBlur, getItemId, onItemClick, }, ref) {
    const mobile = (0, mobile_1.useMobile)();
    const uniqId = (0, hooks_1.useUniqId)();
    const treeSelectId = id !== null && id !== void 0 ? id : uniqId;
    const popupId = `tree-select-popup-${treeSelectId}`;
    const controlWrapRef = react_1.default.useRef(null);
    const controlRef = react_1.default.useRef(null);
    const containerRefLocal = react_1.default.useRef(null);
    const containerRef = propsContainerRef !== null && propsContainerRef !== void 0 ? propsContainerRef : containerRefLocal;
    const handleControlRef = (0, hooks_1.useForkRef)(ref, controlRef);
    const { toggleOpen, open } = (0, useOpenState_1.useOpenState)({
        defaultOpen,
        onClose,
        onOpenChange,
        open: propsOpen,
    });
    const { value, selectedById, setSelected } = (0, useControlledValue_1.useControlledValue)({
        value: propsValue,
        defaultValue,
        onUpdate,
    });
    const list = (0, useList_1.useList)({
        controlledState: {
            selectedById,
            setSelected,
        },
        items,
        getItemId,
        defaultExpandedState,
        withExpandedState,
    });
    const handleItemClick = react_1.default.useMemo(() => {
        if (onItemClick === null) {
            return undefined;
        }
        const handler = (arg, e) => {
            const payload = { id: arg.id, list };
            if (onItemClick) {
                onItemClick === null || onItemClick === void 0 ? void 0 : onItemClick(payload, e);
            }
            else {
                const baseOnClick = (0, useList_1.getListItemClickHandler)({ list, multiple });
                baseOnClick(payload, e);
                const isGroup = list.state.expandedById && arg.id in list.state.expandedById;
                if (!multiple && !isGroup) {
                    toggleOpen(false);
                }
            }
        };
        return handler;
    }, [onItemClick, list, multiple, toggleOpen]);
    // restoring focus when popup opens
    react_1.default.useLayoutEffect(() => {
        var _a;
        if (open) {
            // for some reason popup position on page may be wrong calculated. `preventScroll` prevent page gap in that cases
            (_a = containerRef.current) === null || _a === void 0 ? void 0 : _a.focus({ preventScroll: true });
        }
        return () => list.state.setActiveItemId(undefined); // reset active item on popup close
        // subscribe only in open event
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [open]);
    const handleClose = react_1.default.useCallback(() => toggleOpen(false), [toggleOpen]);
    const { focusWithinProps } = (0, hooks_1.useFocusWithin)({
        onFocusWithin: onFocus,
        onBlurWithin: react_1.default.useCallback((e) => {
            onBlur === null || onBlur === void 0 ? void 0 : onBlur(e);
            handleClose();
        }, [handleClose, onBlur]),
    });
    const controlProps = {
        list,
        open,
        placeholder,
        toggleOpen,
        clearValue: () => list.state.setSelected({}),
        ref: handleControlRef,
        size,
        value,
        disabled,
        id: treeSelectId,
        activeItemId: list.state.activeItemId,
        title,
    };
    const togglerNode = renderControl ? (renderControl(controlProps)) : (react_1.default.createElement(components_1.SelectControl, Object.assign({}, controlProps, { selectedOptionsContent: react_1.default.Children.toArray(value.map((itemId) => itemId in list.structure.itemsById
            ? mapItemDataToContentProps(list.structure.itemsById[itemId]).title
            : '')).join(', '), view: "normal", pin: "round-round", popupId: popupId, selectId: treeSelectId })));
    const mods = Object.assign({}, (width === 'max' && { width }));
    const inlineStyles = {};
    if (typeof width === 'number') {
        inlineStyles.width = width;
    }
    return (react_1.default.createElement("div", Object.assign({ ref: controlWrapRef }, focusWithinProps, { className: b(mods, className), style: inlineStyles }),
        togglerNode,
        react_1.default.createElement(SelectPopup_1.SelectPopup, { ref: controlWrapRef, className: b('popup', { size }, popupClassName), controlRef: controlRef, width: popupWidth, placement: placement, open: open, handleClose: handleClose, disablePortal: popupDisablePortal, mobile: mobile, id: popupId },
            slotBeforeListBody,
            react_1.default.createElement(TreeList_1.TreeList, { list: list, size: size, className: b('list', containerClassName), qa: qa, multiple: multiple, id: `list-${treeSelectId}`, containerRef: containerRef, onItemClick: handleItemClick, renderContainer: renderContainer, mapItemDataToContentProps: mapItemDataToContentProps, renderItem: renderItem !== null && renderItem !== void 0 ? renderItem : defaultItemRenderer }),
            slotAfterListBody)));
});
