// @gravity-ui/uikit/build/cjs/components/Select/components/SelectControl/SelectControl.js
"use strict";
'use client';
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectControl = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const icons_1 = require("@gravity-ui/icons");
const isEmpty_1 = tslib_1.__importDefault(require("lodash/isEmpty"));
const hooks_1 = require("../../../../hooks");
const Icon_1 = require("../../../Icon");
const Popover_1 = require("../../../Popover");
const constants_1 = require("../../constants");
const i18n_1 = tslib_1.__importDefault(require("../../i18n"));
const SelectClear_1 = require("../SelectClear/SelectClear");
const SelectCounter_1 = require("../SelectCounter/SelectCounter");
require("./SelectControl.css");
exports.SelectControl = react_1.default.forwardRef((props, ref) => {
    const { toggleOpen, clearValue, onKeyDown, renderControl, view, size, pin, selectedOptionsContent, className, qa, label, placeholder, isErrorVisible, errorMessage, open, disabled, value, hasClear, popupId, selectId, activeIndex, renderCounter, hasCounter, title, } = props;
    const showOptionsText = Boolean(selectedOptionsContent);
    const showPlaceholder = Boolean(placeholder && !showOptionsText);
    const hasValue = Array.isArray(value) && !(0, isEmpty_1.default)(value.filter(Boolean));
    const errorTooltipId = (0, hooks_1.useUniqId)();
    const [isDisabledButtonAnimation, setIsDisabledButtonAnimation] = react_1.default.useState(false);
    const controlMods = {
        open,
        size,
        pin,
        disabled,
        error: isErrorVisible,
        'has-clear': hasClear,
        'no-active': isDisabledButtonAnimation,
        'has-value': hasValue,
    };
    const buttonMods = {
        open,
        size,
        view,
        pin,
        disabled,
        error: isErrorVisible,
    };
    const handleControlClick = react_1.default.useCallback((e) => {
        // Fix for Safari
        // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#clicking_and_focus
        if (e && e.currentTarget !== document.activeElement && 'focus' in e.currentTarget) {
            e.currentTarget.focus();
        }
        toggleOpen();
    }, [toggleOpen]);
    const disableButtonAnimation = react_1.default.useCallback(() => {
        setIsDisabledButtonAnimation(true);
    }, []);
    const enableButtonAnimation = react_1.default.useCallback(() => {
        setIsDisabledButtonAnimation(false);
    }, []);
    const handleOnClearIconClick = react_1.default.useCallback(() => {
        // return animation on clear click
        setIsDisabledButtonAnimation(false);
        clearValue();
    }, [clearValue]);
    const renderCounterComponent = () => {
        if (!hasCounter) {
            return null;
        }
        const count = value.length;
        const counterComponent = react_1.default.createElement(SelectCounter_1.SelectCounter, { count: count, size: size, disabled: disabled });
        return renderCounter
            ? renderCounter(counterComponent, { count, size, disabled })
            : counterComponent;
    };
    const renderClearIcon = (args) => {
        const valueIsEmpty = value.length === 0;
        if (!hasClear || valueIsEmpty || disabled) {
            return null;
        }
        return (react_1.default.createElement(SelectClear_1.SelectClear, { size: size, onClick: handleOnClearIconClick, onMouseEnter: disableButtonAnimation, onMouseLeave: enableButtonAnimation, renderIcon: args.renderIcon }));
    };
    const triggerProps = {
        id: selectId,
        role: 'combobox',
        'aria-controls': open ? popupId : undefined,
        'aria-haspopup': 'listbox',
        'aria-expanded': open,
        'aria-activedescendant': activeIndex === undefined ? undefined : `${popupId}-item-${activeIndex}`,
        onClick: handleControlClick,
        onKeyDown,
        disabled,
    };
    if (renderControl) {
        return renderControl({
            onKeyDown,
            onClear: clearValue,
            onClick: handleControlClick,
            renderClear: renderClearIcon,
            renderCounter: renderCounterComponent,
            ref,
            open,
            popupId,
            selectId,
            activeIndex,
            disabled,
            triggerProps,
        }, { value });
    }
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement("div", { className: (0, constants_1.selectControlBlock)(controlMods), role: "group" },
            react_1.default.createElement("button", Object.assign({ ref: ref, className: (0, constants_1.selectControlButtonBlock)(buttonMods, className), type: "button", "data-qa": qa, title: title, tabIndex: 0 }, triggerProps),
                label && react_1.default.createElement("span", { className: (0, constants_1.selectControlBlock)('label') }, label),
                showPlaceholder && (react_1.default.createElement("span", { className: (0, constants_1.selectControlBlock)('placeholder') }, placeholder)),
                showOptionsText && (react_1.default.createElement("span", { className: (0, constants_1.selectControlBlock)('option-text') }, selectedOptionsContent))),
            renderCounterComponent(),
            renderClearIcon({}),
            errorMessage && (react_1.default.createElement(Popover_1.Popover, { content: errorMessage, tooltipId: errorTooltipId },
                react_1.default.createElement("button", { "aria-label": (0, i18n_1.default)('label_show-error-info'), "aria-describedby": errorTooltipId, className: (0, constants_1.selectControlBlock)('error-icon') },
                    react_1.default.createElement(Icon_1.Icon, { data: icons_1.TriangleExclamation, size: size === 's' ? 12 : 16 })))),
            react_1.default.createElement(Icon_1.Icon, { className: (0, constants_1.selectControlBlock)('chevron-icon', { disabled }), data: icons_1.ChevronDown, "aria-hidden": "true" }))));
});
exports.SelectControl.displayName = 'SelectControl';
