// @gravity-ui/uikit/build/cjs/hooks/private/useRadioGroup/useRadioGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRadioGroup = void 0;
const __1 = require("../..");
const filterDOMProps_1 = require("../../../components/utils/filterDOMProps");
const useFormResetHandler_1 = require("../useFormResetHandler");
function useRadioGroup(props) {
    var _a, _b;
    const { name, value, defaultValue, options = [], disabled, onUpdate, onChange, onFocus, onBlur, } = props;
    const controlId = (0, __1.useUniqId)();
    const [currentValue, setValueState] = (0, __1.useControlledState)(value, 
    // FIXME: Do not set defaultValue to first option value
    (_b = defaultValue !== null && defaultValue !== void 0 ? defaultValue : (_a = options[0]) === null || _a === void 0 ? void 0 : _a.value) !== null && _b !== void 0 ? _b : null, onUpdate);
    const fieldRef = (0, useFormResetHandler_1.useFormResetHandler)({
        initialValue: currentValue,
        onReset: setValueState,
    });
    const handleChange = (event) => {
        setValueState(event.target.value);
        if (onChange) {
            onChange(event);
        }
    };
    const containerProps = Object.assign(Object.assign({}, (0, filterDOMProps_1.filterDOMProps)(props, { labelable: true })), { role: 'radiogroup', 'aria-disabled': disabled });
    const optionsProps = options.map((option) => ({
        name: name || controlId,
        value: option.value,
        content: option.content,
        title: option.title,
        checked: currentValue === String(option.value),
        disabled: disabled || option.disabled,
        onChange: handleChange,
        // FIXME: onFocus and onBlur should be on the container via useFocusWithin hook
        onFocus: onFocus,
        onBlur: onBlur,
        ref: fieldRef,
    }));
    return { containerProps, optionsProps };
}
exports.useRadioGroup = useRadioGroup;
