// @gravity-ui/uikit/build/esm/components/Link/Link.js
'use client';
import React from 'react';
import { block } from '../utils/cn';
import { eventBroker } from '../utils/event-broker';
import './Link.css';
const b = block('link');
export const Link = React.forwardRef(function Link({ view = 'normal', visitable, underline, href, target, rel, title, children, extraProps, onClick, onFocus, onBlur, id, style, className, qa, }, ref) {
    const handleClickCapture = React.useCallback((event) => {
        eventBroker.publish({
            componentId: 'Link',
            eventId: 'click',
            domEvent: event,
        });
    }, []);
    const commonProps = {
        title,
        onClick,
        onClickCapture: handleClickCapture,
        onFocus,
        onBlur,
        id,
        style,
        className: b({ view, visitable, underline }, className),
        'data-qa': qa,
    };
    const relProp = target === '_blank' && !rel ? 'noopener noreferrer' : rel;
    return (React.createElement("a", Object.assign({}, extraProps, commonProps, { ref: ref, href: href, target: target, rel: relProp }), children));
});
