// @gravity-ui/uikit/build/esm/components/TreeList/TreeList.js
'use client';
import React from 'react';
import { useUniqId } from '../../hooks';
import { ListContainer, ListItemView, getItemRenderState, getListItemClickHandler, useListKeydown, } from '../useList';
import { block } from '../utils/cn';
const b = block('tree-list');
export const TreeList = ({ qa, id, size = 'm', className, list, multiple, containerRef: propsContainerRef, renderItem: propsRenderItem, renderContainer = ListContainer, onItemClick: propsOnItemClick, mapItemDataToContentProps, }) => {
    const uniqId = useUniqId();
    const treeListId = id !== null && id !== void 0 ? id : uniqId;
    const containerRefLocal = React.useRef(null);
    const containerRef = propsContainerRef !== null && propsContainerRef !== void 0 ? propsContainerRef : containerRefLocal;
    const onItemClick = React.useMemo(() => {
        if (propsOnItemClick === null) {
            return undefined;
        }
        const handler = (arg, e) => {
            const payload = { id: arg.id, list };
            if (propsOnItemClick) {
                propsOnItemClick === null || propsOnItemClick === void 0 ? void 0 : propsOnItemClick(payload, e);
            }
            else {
                const baseOnClick = getListItemClickHandler({ list, multiple });
                baseOnClick(payload, e);
            }
        };
        return handler;
    }, [propsOnItemClick, list, multiple]);
    useListKeydown({
        containerRef,
        onItemClick,
        list,
    });
    const renderItem = (itemId, index, renderContainerProps) => {
        const renderState = getItemRenderState({
            qa,
            id: itemId,
            size,
            multiple,
            mapItemDataToContentProps,
            onItemClick,
            list,
        });
        if (propsRenderItem) {
            return propsRenderItem({
                id: itemId,
                data: renderState.data,
                props: renderState.props,
                context: renderState.context,
                index,
                renderContainerProps,
                list,
            });
        }
        return React.createElement(ListItemView, Object.assign({}, renderState.props, renderContainerProps));
    };
    // not JSX decl here is from weird `react-beautiful-dnd` render bug
    return renderContainer({
        qa,
        id: `list-${treeListId}`,
        size,
        containerRef,
        className: b(null, className),
        list,
        renderItem,
    });
};
