// @gravity-ui/uikit/build/esm/components/Button/Button.js
'use client';
import React from 'react';
import { block } from '../utils/cn';
import { isIcon, isSvg } from '../utils/common';
import { eventBroker } from '../utils/event-broker';
import { isOfType } from '../utils/isOfType';
import { ButtonIcon, getIconSide } from './ButtonIcon';
import './Button.css';
const b = block('button');
const ButtonWithHandlers = React.forwardRef(function Button({ view = 'normal', size = 'm', pin = 'round-round', selected, disabled = false, loading = false, width, title, tabIndex, type = 'button', component, href, target, rel, extraProps, onClick, onMouseEnter, onMouseLeave, onFocus, onBlur, children, id, style, className, qa, }, ref) {
    const handleClickCapture = React.useCallback((event) => {
        eventBroker.publish({
            componentId: 'Button',
            eventId: 'click',
            domEvent: event,
            meta: {
                content: event.currentTarget.textContent,
                view,
            },
        });
    }, [view]);
    const commonProps = {
        title,
        tabIndex,
        onClick,
        onClickCapture: handleClickCapture,
        onMouseEnter,
        onMouseLeave,
        onFocus,
        onBlur,
        id,
        style,
        className: b({
            view,
            size,
            pin,
            selected,
            disabled: disabled || loading,
            loading,
            width,
        }, className),
        'data-qa': qa,
    };
    if (typeof href === 'string' || component) {
        const linkProps = {
            href,
            target,
            rel: target === '_blank' && !rel ? 'noopener noreferrer' : rel,
        };
        return React.createElement(component || 'a', Object.assign(Object.assign(Object.assign(Object.assign({}, extraProps), commonProps), (component ? {} : linkProps)), { ref: ref, 'aria-disabled': disabled || loading }), prepareChildren(children));
    }
    else {
        return (React.createElement("button", Object.assign({}, extraProps, commonProps, { ref: ref, type: type, disabled: disabled || loading, "aria-pressed": selected }), prepareChildren(children)));
    }
});
ButtonWithHandlers.displayName = 'Button';
export const Button = Object.assign(ButtonWithHandlers, { Icon: ButtonIcon });
const isButtonIconComponent = isOfType(ButtonIcon);
const isSpan = isOfType('span');
const buttonIconClassRe = RegExp(`^${b('icon')}($|\\s+\\w)`);
// eslint-disable-next-line complexity
function prepareChildren(children) {
    const items = React.Children.toArray(children);
    if (items.length === 1) {
        const onlyItem = items[0];
        const isButtonIconElement = isButtonIconComponent(onlyItem) ||
            (isSpan(onlyItem) && buttonIconClassRe.test(onlyItem.props.className || ''));
        if (isButtonIconElement) {
            return onlyItem;
        }
        else if (isIcon(onlyItem) || isSvg(onlyItem)) {
            return React.createElement(Button.Icon, { key: "icon" }, onlyItem);
        }
        else {
            return (React.createElement("span", { key: "text", className: b('text') }, onlyItem));
        }
    }
    else {
        let startIcon, endIcon, text;
        const content = [];
        for (const item of items) {
            const isIconElement = isIcon(item) || isSvg(item);
            const isButtonIconElement = isButtonIconComponent(item);
            const isRenderedButtonIconElement = isSpan(item) && buttonIconClassRe.test(item.props.className || '');
            if (isIconElement || isButtonIconElement || isRenderedButtonIconElement) {
                if (!startIcon && content.length === 0) {
                    const key = 'icon-start';
                    const side = 'start';
                    if (isIconElement) {
                        startIcon = (React.createElement(Button.Icon, { key: key, side: side }, item));
                    }
                    else if (isButtonIconElement) {
                        startIcon = React.cloneElement(item, {
                            side,
                        });
                    }
                    else {
                        startIcon = React.cloneElement(item, {
                            className: b('icon', { side: getIconSide(side) }, item.props.className),
                        });
                    }
                }
                else if (!endIcon && content.length !== 0) {
                    const key = 'icon-end';
                    const side = 'end';
                    if (isIconElement) {
                        endIcon = (React.createElement(Button.Icon, { key: key, side: side }, item));
                    }
                    else if (isButtonIconElement) {
                        endIcon = React.cloneElement(item, {
                            side,
                        });
                    }
                    else {
                        endIcon = React.cloneElement(item, {
                            className: b('icon', { side: getIconSide(side) }, item.props.className),
                        });
                    }
                }
            }
            else {
                content.push(item);
            }
        }
        if (content.length > 0) {
            text = (React.createElement("span", { key: "text", className: b('text') }, content));
        }
        return [startIcon, endIcon, text];
    }
}
