// @gravity-ui/uikit/build/cjs/components/useList/hooks/useList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useList = void 0;
const tslib_1 = require("tslib");
/* eslint-disable valid-jsdoc */
const react_1 = tslib_1.__importDefault(require("react"));
const useFlattenListItems_1 = require("./useFlattenListItems");
const useListParsedState_1 = require("./useListParsedState");
const useListState_1 = require("./useListState");
/**
 * Take array of items as a argument with params described what type of list initial data represents.
 */
const useList = ({ items, getItemId, defaultExpandedState = 'expanded', withExpandedState = true, initialState: initialValues, controlledState, }) => {
    const { itemsById, groupsState, itemsState, initialState } = (0, useListParsedState_1.useListParsedState)({
        items,
        getItemId,
        defaultExpandedState,
    });
    const initValues = react_1.default.useMemo(() => {
        return {
            expandedById: Object.assign(Object.assign({}, initialState.expandedById), initialValues === null || initialValues === void 0 ? void 0 : initialValues.expandedById),
            selectedById: Object.assign(Object.assign({}, initialState.selectedById), initialValues === null || initialValues === void 0 ? void 0 : initialValues.selectedById),
            disabledById: Object.assign(Object.assign({}, initialState.disabledById), initialValues === null || initialValues === void 0 ? void 0 : initialValues.disabledById),
            activeItemId: initialValues === null || initialValues === void 0 ? void 0 : initialValues.activeItemId,
        };
    }, [
        initialState.disabledById,
        initialState.expandedById,
        initialState.selectedById,
        initialValues === null || initialValues === void 0 ? void 0 : initialValues.activeItemId,
        initialValues === null || initialValues === void 0 ? void 0 : initialValues.disabledById,
        initialValues === null || initialValues === void 0 ? void 0 : initialValues.expandedById,
        initialValues === null || initialValues === void 0 ? void 0 : initialValues.selectedById,
    ]);
    const innerState = (0, useListState_1.useListState)({
        initialState: initValues,
        withExpandedState,
    });
    const realState = react_1.default.useMemo(() => {
        if (controlledState) {
            return Object.assign(Object.assign({}, innerState), controlledState);
        }
        return innerState;
    }, [controlledState, innerState]);
    const { visibleFlattenIds, idToFlattenIndex, rootIds } = (0, useFlattenListItems_1.useFlattenListItems)({
        items,
        /**
         * By default controlled from list items declaration state
         */
        expandedById: realState.expandedById,
        getItemId,
    });
    return {
        state: realState,
        structure: {
            rootIds,
            items,
            visibleFlattenIds,
            idToFlattenIndex,
            itemsById,
            groupsState,
            itemsState,
        },
    };
};
exports.useList = useList;
