// @gravity-ui/uikit/build/esm/components/List/components/ListItem.js
'use client';
import React from 'react';
import { Grip } from '@gravity-ui/icons';
import { Icon } from '../../Icon';
import { block } from '../../utils/cn';
import { eventBroker } from '../../utils/event-broker';
import { ListQa } from '../constants';
const b = block('list');
export const defaultRenderItem = (item) => String(item);
function getStyle(provided, style) {
    if (!style) {
        return provided === null || provided === void 0 ? void 0 : provided.draggableProps.style;
    }
    return Object.assign(Object.assign({}, provided === null || provided === void 0 ? void 0 : provided.draggableProps.style), style);
}
export class ListItem extends React.Component {
    constructor() {
        super(...arguments);
        this.node = null;
        this.getNode = () => this.node;
        this.setRef = (node) => {
            var _a;
            this.node = node;
            (_a = this.props.provided) === null || _a === void 0 ? void 0 : _a.innerRef(node);
        };
        this.onClick = (event) => {
            if (!this.props.onClick) {
                return;
            }
            this.props.onClick(this.props.item, this.props.itemIndex, false, event);
        };
        this.onClickCapture = (event) => {
            ListItem.publishEvent({
                domEvent: event,
                eventId: 'click',
            });
        };
        this.onMouseEnter = () => !this.props.item.disabled && this.props.onActivate(this.props.itemIndex);
    }
    render() {
        var _a, _b;
        const { item, height, style, sortable, sortHandleAlign, itemClassName, selected, active, role = 'listitem', isDragging = false, } = this.props;
        /*
        This fixes item drag layout for rtl direction.
        react-window has a bug where in rtl it setting "right" to 0 instead of undefined.
         */
        const fixedStyle = Object.assign(Object.assign({ height }, style), { right: undefined });
        return (
        // eslint-disable-next-line jsx-a11y/click-events-have-key-events
        React.createElement("div", Object.assign({ role: role, "aria-selected": selected, "aria-disabled": item.disabled, "data-qa": active ? ListQa.ACTIVE_ITEM : undefined, className: b('item', {
                sortable,
                active,
                selected,
                inactive: item.disabled,
                'sort-handle-align': sortHandleAlign,
                dragging: isDragging,
            }, itemClassName) }, (_a = this.props.provided) === null || _a === void 0 ? void 0 : _a.draggableProps, (_b = this.props.provided) === null || _b === void 0 ? void 0 : _b.dragHandleProps, { style: getStyle(this.props.provided, fixedStyle), onClick: item.disabled ? undefined : this.onClick, onClickCapture: item.disabled ? undefined : this.onClickCapture, onMouseEnter: this.onMouseEnter, ref: this.setRef, id: `${this.props.listId}-item-${this.props.itemIndex}` }),
            this.renderSortIcon(),
            this.renderContent()));
    }
    renderSortIcon() {
        const { sortable } = this.props;
        return sortable ? (React.createElement("div", { className: b('item-sort-icon') },
            React.createElement(Icon, { data: Grip, size: 12 }))) : null;
    }
    renderContent() {
        const { renderItem = defaultRenderItem, item, active, itemIndex } = this.props;
        return React.createElement("div", { className: b('item-content') }, renderItem(item, active, itemIndex));
    }
}
ListItem.publishEvent = eventBroker.withEventPublisher('List');
