// @gravity-ui/uikit/build/esm/components/List/components/SimpleContainer.js
'use client';
import React from 'react';
import _range from 'lodash/range';
function getRefs(count) {
    return _range(count).reduce((acc, index) => {
        acc[index] = React.createRef();
        return acc;
    }, {});
}
export class SimpleContainer extends React.Component {
    static getDerivedStateFromProps({ itemCount }, prevState) {
        const refsCount = Object.keys(prevState.refsList).length;
        if (itemCount === refsCount) {
            return prevState;
        }
        else {
            return {
                refsList: getRefs(itemCount),
            };
        }
    }
    constructor(props) {
        super(props);
        this.node = null;
        this.setRef = (node) => {
            var _a;
            this.node = node;
            (_a = this.props.provided) === null || _a === void 0 ? void 0 : _a.innerRef(node);
        };
        this.state = {
            refsList: getRefs(props.itemCount),
        };
    }
    render() {
        const children = React.Children.map(this.props.children, (child, index) => React.cloneElement(child, { ref: this.state.refsList[index] }));
        return React.createElement("div", { ref: this.setRef }, children);
    }
    scrollToItem(index) {
        var _a, _b, _c, _d;
        const listItem = (_a = this.state.refsList[index]) === null || _a === void 0 ? void 0 : _a.current;
        if (listItem && typeof listItem.getNode === 'function') {
            const node = listItem.getNode();
            if (node) {
                if (!((_c = (_b = this.props).onScrollToItem) === null || _c === void 0 ? void 0 : _c.call(_b, node))) {
                    (_d = node.scrollIntoView) === null || _d === void 0 ? void 0 : _d.call(node, { block: 'nearest' });
                }
            }
        }
    }
}
