// @gravity-ui/uikit/build/cjs/components/Select/components/SelectFilter/SelectFilter.js
"use strict";
'use client';
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectFilter = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const controls_1 = require("../../../controls");
const cn_1 = require("../../../utils/cn");
const constants_1 = require("../../constants");
const i18n_1 = tslib_1.__importDefault(require("../../i18n"));
require("./SelectFilter.css");
const b = (0, cn_1.block)('select-filter');
const style = {
    padding: '4px 4px 0',
};
exports.SelectFilter = react_1.default.forwardRef((props, ref) => {
    const { onChange, onKeyDown, renderFilter, size, value, placeholder, popupId, activeIndex } = props;
    const inputRef = react_1.default.useRef(null);
    react_1.default.useImperativeHandle(ref, () => ({
        focus: () => { var _a; return (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus({ preventScroll: true }); },
    }), []);
    const inputProps = {
        value,
        placeholder,
        size: 1,
        onKeyDown,
        onChange: (e) => {
            onChange(e.target.value);
        },
        'aria-label': (0, i18n_1.default)('label_filter'),
        'aria-controls': popupId,
        'aria-activedescendant': activeIndex === undefined ? undefined : `${popupId}-item-${activeIndex}`,
    };
    if (renderFilter) {
        return renderFilter({ onChange, onKeyDown, value, ref: inputRef, style, inputProps });
    }
    return (react_1.default.createElement("div", { className: b(), style: style },
        react_1.default.createElement(controls_1.TextInput, { controlRef: inputRef, controlProps: {
                className: b('input'),
                size: 1,
                'aria-label': inputProps['aria-label'],
                'aria-controls': inputProps['aria-controls'],
                'aria-activedescendant': inputProps['aria-activedescendant'],
            }, size: size, value: value, placeholder: placeholder, onUpdate: onChange, onKeyDown: onKeyDown, qa: constants_1.SelectQa.FILTER_INPUT })));
});
exports.SelectFilter.displayName = 'SelectFilter';
