// @gravity-ui/uikit/build/esm/components/TreeSelect/TreeSelect.js
'use client';
import React from 'react';
import { useFocusWithin, useForkRef, useUniqId } from '../../hooks';
import { useOpenState } from '../../hooks/useSelect/useOpenState';
import { SelectControl } from '../Select/components';
import { SelectPopup } from '../Select/components/SelectPopup/SelectPopup';
import { TreeList } from '../TreeList';
import { useMobile } from '../mobile';
import { ListItemView, getListItemClickHandler, useList } from '../useList';
import { block } from '../utils/cn';
import { useControlledValue } from './hooks/useControlledValue';
import './TreeSelect.css';
const b = block('tree-select');
const defaultItemRenderer = (renderState) => {
    return React.createElement(ListItemView, Object.assign({}, renderState.props, renderState.renderContainerProps));
};
export const TreeSelect = React.forwardRef(function TreeSelect({ id, qa, title, placement, slotBeforeListBody, slotAfterListBody, size = 'm', defaultOpen, width, containerRef: propsContainerRef, className, containerClassName, popupClassName, open: propsOpen, multiple, popupWidth, popupDisablePortal, items, value: propsValue, defaultValue, placeholder, disabled = false, withExpandedState = true, defaultExpandedState = 'expanded', onClose, onOpenChange, onUpdate, renderControl, renderItem = defaultItemRenderer, renderContainer, mapItemDataToContentProps, onFocus, onBlur, getItemId, onItemClick, }, ref) {
    const mobile = useMobile();
    const uniqId = useUniqId();
    const treeSelectId = id !== null && id !== void 0 ? id : uniqId;
    const popupId = `tree-select-popup-${treeSelectId}`;
    const controlWrapRef = React.useRef(null);
    const controlRef = React.useRef(null);
    const containerRefLocal = React.useRef(null);
    const containerRef = propsContainerRef !== null && propsContainerRef !== void 0 ? propsContainerRef : containerRefLocal;
    const handleControlRef = useForkRef(ref, controlRef);
    const { toggleOpen, open } = useOpenState({
        defaultOpen,
        onClose,
        onOpenChange,
        open: propsOpen,
    });
    const { value, selectedById, setSelected } = useControlledValue({
        value: propsValue,
        defaultValue,
        onUpdate,
    });
    const list = useList({
        controlledState: {
            selectedById,
            setSelected,
        },
        items,
        getItemId,
        defaultExpandedState,
        withExpandedState,
    });
    const handleItemClick = React.useMemo(() => {
        if (onItemClick === null) {
            return undefined;
        }
        const handler = (arg, e) => {
            const payload = { id: arg.id, list };
            if (onItemClick) {
                onItemClick === null || onItemClick === void 0 ? void 0 : onItemClick(payload, e);
            }
            else {
                const baseOnClick = getListItemClickHandler({ list, multiple });
                baseOnClick(payload, e);
                const isGroup = list.state.expandedById && arg.id in list.state.expandedById;
                if (!multiple && !isGroup) {
                    toggleOpen(false);
                }
            }
        };
        return handler;
    }, [onItemClick, list, multiple, toggleOpen]);
    // restoring focus when popup opens
    React.useLayoutEffect(() => {
        var _a;
        if (open) {
            // for some reason popup position on page may be wrong calculated. `preventScroll` prevent page gap in that cases
            (_a = containerRef.current) === null || _a === void 0 ? void 0 : _a.focus({ preventScroll: true });
        }
        return () => list.state.setActiveItemId(undefined); // reset active item on popup close
        // subscribe only in open event
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [open]);
    const handleClose = React.useCallback(() => toggleOpen(false), [toggleOpen]);
    const { focusWithinProps } = useFocusWithin({
        onFocusWithin: onFocus,
        onBlurWithin: React.useCallback((e) => {
            onBlur === null || onBlur === void 0 ? void 0 : onBlur(e);
            handleClose();
        }, [handleClose, onBlur]),
    });
    const controlProps = {
        list,
        open,
        placeholder,
        toggleOpen,
        clearValue: () => list.state.setSelected({}),
        ref: handleControlRef,
        size,
        value,
        disabled,
        id: treeSelectId,
        activeItemId: list.state.activeItemId,
        title,
    };
    const togglerNode = renderControl ? (renderControl(controlProps)) : (React.createElement(SelectControl, Object.assign({}, controlProps, { selectedOptionsContent: React.Children.toArray(value.map((itemId) => itemId in list.structure.itemsById
            ? mapItemDataToContentProps(list.structure.itemsById[itemId]).title
            : '')).join(', '), view: "normal", pin: "round-round", popupId: popupId, selectId: treeSelectId })));
    const mods = Object.assign({}, (width === 'max' && { width }));
    const inlineStyles = {};
    if (typeof width === 'number') {
        inlineStyles.width = width;
    }
    return (React.createElement("div", Object.assign({ ref: controlWrapRef }, focusWithinProps, { className: b(mods, className), style: inlineStyles }),
        togglerNode,
        React.createElement(SelectPopup, { ref: controlWrapRef, className: b('popup', { size }, popupClassName), controlRef: controlRef, width: popupWidth, placement: placement, open: open, handleClose: handleClose, disablePortal: popupDisablePortal, mobile: mobile, id: popupId },
            slotBeforeListBody,
            React.createElement(TreeList, { list: list, size: size, className: b('list', containerClassName), qa: qa, multiple: multiple, id: `list-${treeSelectId}`, containerRef: containerRef, onItemClick: handleItemClick, renderContainer: renderContainer, mapItemDataToContentProps: mapItemDataToContentProps, renderItem: renderItem !== null && renderItem !== void 0 ? renderItem : defaultItemRenderer }),
            slotAfterListBody)));
});
