// @gravity-ui/uikit/build/esm/components/controls/TextArea/TextArea.js
'use client';
import React from 'react';
import { useControlledState, useForkRef, useUniqId } from '../../../hooks';
import { useFormResetHandler } from '../../../hooks/private';
import { block } from '../../utils/cn';
import { ClearButton, mapTextInputSizeToButtonSize } from '../common';
import { OuterAdditionalContent } from '../common/OuterAdditionalContent/OuterAdditionalContent';
import { errorPropsMapper, getInputControlState, prepareAutoComplete } from '../utils';
import { TextAreaControl } from './TextAreaControl';
import './TextArea.css';
const b = block('text-area');
// eslint-disable-next-line complexity
export const TextArea = React.forwardRef(function TextArea(props, ref) {
    const { view = 'normal', size = 'm', pin = 'round-round', name, value, defaultValue, disabled, readOnly, hasClear = false, error, errorMessage: errorMessageProp, validationState: validationStateProp, autoComplete, id: idProp, tabIndex, style, className, qa, controlProps, note, onUpdate, onChange, } = props;
    const { errorMessage, validationState } = errorPropsMapper({
        error,
        errorMessage: errorMessageProp,
        validationState: validationStateProp,
    });
    const [inputValue, setInputValue] = useControlledState(value, defaultValue !== null && defaultValue !== void 0 ? defaultValue : '', onUpdate);
    const innerControlRef = React.useRef(null);
    const fieldRef = useFormResetHandler({ initialValue: inputValue, onReset: setInputValue });
    const handleRef = useForkRef(props.controlRef, innerControlRef, fieldRef);
    const [hasVerticalScrollbar, setHasVerticalScrollbar] = React.useState(false);
    const state = getInputControlState(validationState);
    const innerId = useUniqId();
    const isErrorMsgVisible = validationState === 'invalid' && Boolean(errorMessage);
    const isClearControlVisible = Boolean(hasClear && !disabled && !readOnly && inputValue);
    const id = idProp || innerId;
    const errorMessageId = useUniqId();
    const noteId = useUniqId();
    const ariaDescribedBy = [
        controlProps === null || controlProps === void 0 ? void 0 : controlProps['aria-describedby'],
        note ? noteId : undefined,
        isErrorMsgVisible ? errorMessageId : undefined,
    ]
        .filter(Boolean)
        .join(' ');
    const commonProps = {
        id,
        tabIndex,
        name,
        onChange(event) {
            setInputValue(event.target.value);
            if (onChange) {
                onChange(event);
            }
        },
        autoComplete: prepareAutoComplete(autoComplete),
        controlProps: Object.assign(Object.assign({}, controlProps), { 'aria-describedby': ariaDescribedBy || undefined, 'aria-invalid': validationState === 'invalid' || undefined }),
    };
    const handleClear = (event) => {
        const control = innerControlRef.current;
        if (control) {
            control.focus();
            const syntheticEvent = Object.create(event);
            syntheticEvent.target = control;
            syntheticEvent.currentTarget = control;
            control.value = '';
            if (onChange) {
                onChange(syntheticEvent);
            }
        }
        setInputValue('');
    };
    React.useEffect(() => {
        const control = innerControlRef.current;
        if (control) {
            const currHasVerticalScrollbar = control.scrollHeight > control.clientHeight;
            if (hasVerticalScrollbar !== currHasVerticalScrollbar) {
                setHasVerticalScrollbar(currHasVerticalScrollbar);
            }
        }
    }, [inputValue, hasVerticalScrollbar]);
    return (React.createElement("span", { ref: ref, style: style, className: b({
            view,
            size,
            disabled,
            state,
            pin: view === 'clear' ? undefined : pin,
            'has-clear': isClearControlVisible,
            'has-scrollbar': hasVerticalScrollbar,
        }, className), "data-qa": qa },
        React.createElement("span", { className: b('content') },
            React.createElement(TextAreaControl, Object.assign({}, props, commonProps, { controlRef: handleRef })),
            isClearControlVisible && (React.createElement(ClearButton, { className: b('clear', { size }), size: mapTextInputSizeToButtonSize(size), onClick: handleClear }))),
        React.createElement(OuterAdditionalContent, { errorMessage: isErrorMsgVisible ? errorMessage : null, errorMessageId: errorMessageId, note: note, noteId: noteId })));
});
