// @gravity-ui/uikit/build/cjs/components/Link/Link.js
"use strict";
'use client';
Object.defineProperty(exports, "__esModule", { value: true });
exports.Link = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const cn_1 = require("../utils/cn");
const event_broker_1 = require("../utils/event-broker");
require("./Link.css");
const b = (0, cn_1.block)('link');
exports.Link = react_1.default.forwardRef(function Link({ view = 'normal', visitable, underline, href, target, rel, title, children, extraProps, onClick, onFocus, onBlur, id, style, className, qa, }, ref) {
    const handleClickCapture = react_1.default.useCallback((event) => {
        event_broker_1.eventBroker.publish({
            componentId: 'Link',
            eventId: 'click',
            domEvent: event,
        });
    }, []);
    const commonProps = {
        title,
        onClick,
        onClickCapture: handleClickCapture,
        onFocus,
        onBlur,
        id,
        style,
        className: b({ view, visitable, underline }, className),
        'data-qa': qa,
    };
    const relProp = target === '_blank' && !rel ? 'noopener noreferrer' : rel;
    return (react_1.default.createElement("a", Object.assign({}, extraProps, commonProps, { ref: ref, href: href, target: target, rel: relProp }), children));
});
