// @gravity-ui/uikit/build/cjs/components/Disclosure/DisclosureSummary/DisclosureSummary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultDisclosureSummary = exports.DisclosureSummary = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const ArrowToggle_1 = require("../../ArrowToggle");
const warn_1 = require("../../utils/warn");
const DisclosureContext_1 = require("../DisclosureContext");
const constants_1 = require("../constants");
const ComponentSizeToIconSizeMap = {
    m: 14,
    l: 16,
    xl: 20,
};
function warnAboutPhysicalValues() {
    (0, warn_1.warnOnce)('[Disclosure] Physical values (left, right) of "arrowPosition" property are deprecated. Use logical values (start, end) instead.');
}
function DisclosureSummary({ children: renderFunction }) {
    const handleToggle = (0, DisclosureContext_1.useToggleDisclosure)();
    const { ariaControls, ariaLabelledby: id, expanded, disabled } = (0, DisclosureContext_1.useDisclosureAttributes)();
    const props = { onClick: handleToggle, ariaControls, id, expanded, disabled };
    return renderFunction(props, react_1.default.createElement(DefaultDisclosureSummary, Object.assign({}, props)));
}
exports.DisclosureSummary = DisclosureSummary;
function DefaultDisclosureSummary({ onClick, ariaControls, id, expanded, disabled, }) {
    const { size, summary, arrowPosition } = (0, DisclosureContext_1.useDisclosureAttributes)();
    let arrowMod = arrowPosition;
    if (arrowMod === 'left') {
        warnAboutPhysicalValues();
        arrowMod = 'start';
    }
    if (arrowMod === 'right') {
        warnAboutPhysicalValues();
        arrowMod = 'end';
    }
    return (react_1.default.createElement("button", { type: "button", "aria-expanded": expanded, className: (0, constants_1.b)('trigger', { disabled, arrow: arrowMod }), "aria-controls": ariaControls, id: id, onClick: onClick, disabled: disabled, "data-qa": constants_1.DisclosureQa.SUMMARY },
        react_1.default.createElement(ArrowToggle_1.ArrowToggle, { size: ComponentSizeToIconSizeMap[size], direction: expanded ? 'top' : 'bottom' }),
        summary));
}
exports.DefaultDisclosureSummary = DefaultDisclosureSummary;
DisclosureSummary.displayName = 'DisclosureSummary';
