// @gravity-ui/uikit/build/esm/components/Select/components/SelectFilter/SelectFilter.js
'use client';
import React from 'react';
import { TextInput } from '../../../controls';
import { block } from '../../../utils/cn';
import { SelectQa } from '../../constants';
import i18n from '../../i18n';
import './SelectFilter.css';
const b = block('select-filter');
const style = {
    padding: '4px 4px 0',
};
export const SelectFilter = React.forwardRef((props, ref) => {
    const { onChange, onKeyDown, renderFilter, size, value, placeholder, popupId, activeIndex } = props;
    const inputRef = React.useRef(null);
    React.useImperativeHandle(ref, () => ({
        focus: () => { var _a; return (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus({ preventScroll: true }); },
    }), []);
    const inputProps = {
        value,
        placeholder,
        size: 1,
        onKeyDown,
        onChange: (e) => {
            onChange(e.target.value);
        },
        'aria-label': i18n('label_filter'),
        'aria-controls': popupId,
        'aria-activedescendant': activeIndex === undefined ? undefined : `${popupId}-item-${activeIndex}`,
    };
    if (renderFilter) {
        return renderFilter({ onChange, onKeyDown, value, ref: inputRef, style, inputProps });
    }
    return (React.createElement("div", { className: b(), style: style },
        React.createElement(TextInput, { controlRef: inputRef, controlProps: {
                className: b('input'),
                size: 1,
                'aria-label': inputProps['aria-label'],
                'aria-controls': inputProps['aria-controls'],
                'aria-activedescendant': inputProps['aria-activedescendant'],
            }, size: size, value: value, placeholder: placeholder, onUpdate: onChange, onKeyDown: onKeyDown, qa: SelectQa.FILTER_INPUT })));
});
SelectFilter.displayName = 'SelectFilter';
