// @gravity-ui/uikit/build/esm/components/Slider/HandleWithTooltip/HandleWithTooltip.js
import React from 'react';
import { SliderTooltip } from '../SliderTooltip/SliderTooltip';
export const HandleWithTooltip = ({ originHandle, originHandleProps, stateModifiers, tooltipFormat, className, }) => {
    var _a;
    const autoVisible = stateModifiers['tooltip-display'] === 'auto';
    const alwaysVisible = stateModifiers['tooltip-display'] === 'on';
    const [tooltipVisible, setTooltipVisible] = React.useState(false);
    const [focused, setFocused] = React.useState(false);
    const [hovered, setHovered] = React.useState(false);
    if (alwaysVisible && !tooltipVisible) {
        setTooltipVisible(true);
    }
    //to show tooltip on mobile devices on touch
    if (autoVisible && !tooltipVisible && originHandleProps.dragging) {
        setTooltipVisible(true);
    }
    const styleProp = stateModifiers.rtl ? 'right' : 'left';
    const tooltipContent = tooltipFormat
        ? tooltipFormat(originHandleProps.value)
        : originHandleProps.value;
    const handleTooltipVisibility = ({ currentHovered, currentFocused, }) => {
        if (autoVisible) {
            const handleHovered = currentHovered === undefined ? hovered : currentHovered;
            const handleFocused = currentFocused === undefined ? focused : currentFocused;
            setTooltipVisible(handleHovered || handleFocused);
        }
    };
    const handle = alwaysVisible
        ? originHandle
        : React.cloneElement(originHandle, {
            onMouseEnter: (event) => {
                var _a, _b;
                (_b = (_a = originHandle.props).onMouseEnter) === null || _b === void 0 ? void 0 : _b.call(_a, event);
                setHovered(true);
                handleTooltipVisibility({ currentHovered: true });
            },
            onMouseLeave: (event) => {
                var _a, _b;
                (_b = (_a = originHandle.props).onMouseLeave) === null || _b === void 0 ? void 0 : _b.call(_a, event);
                setHovered(false);
                handleTooltipVisibility({ currentHovered: false });
            },
            onFocus: (event) => {
                var _a, _b;
                (_b = (_a = originHandle.props).onFocus) === null || _b === void 0 ? void 0 : _b.call(_a, event);
                setFocused(true);
                handleTooltipVisibility({ currentFocused: true });
            },
            onBlur: (event) => {
                var _a, _b;
                (_b = (_a = originHandle.props).onBlur) === null || _b === void 0 ? void 0 : _b.call(_a, event);
                setFocused(false);
                handleTooltipVisibility({ currentFocused: false });
            },
        });
    return (React.createElement(React.Fragment, null,
        handle,
        tooltipVisible && (React.createElement(SliderTooltip, { className: className, style: {
                insetInlineStart: (_a = originHandle.props.style) === null || _a === void 0 ? void 0 : _a[styleProp],
            }, stateModifiers: stateModifiers }, tooltipContent))));
};
